xla/service/gpu/tests/gpu_reduce_scatter_all_gather_combiner_test.cc

```diff
@@ -0,0 +1,169 @@
+  }
+
+  size_t AllGatherCount(std::unique_ptr<HloModule> &module) {
+    return absl::c_count_if(module->entry_computation()->instructions(),
+                            [](const HloInstruction *inst) {
+                              return inst->opcode() == HloOpcode::kAllGather;
+                            });
+  }
+  size_t AllReduceCount(std::unique_ptr<HloModule> &module) {
+    return absl::c_count_if(module->entry_computation()->instructions(),
+                            [](const HloInstruction *inst) {
+                              return inst->opcode() == HloOpcode::kAllReduce;
+                            });
+  }
+  size_t ReduceScatterCount(std::unique_ptr<HloModule> &module) {
+    return absl::c_count_if(module->entry_computation()->instructions(),
+                            [](const HloInstruction *inst) {
+                              return inst->opcode() ==
+                                     HloOpcode::kReduceScatter;
+                            });
+  }
+};
+
+TEST_F(ReduceScatterAllGatherCombinerTest, BranchesOptimized) {
+  absl::string_view hlo_string = R"(
+HloModule ReduceScatter
+
+add {
+  x = bf16[] parameter(0)
+  y = bf16[] parameter(1)
+  ROOT add = bf16[] add(x, y)
+}
+
+ENTRY main {
+param.1 = bf16[8,128,1024]{2,1,0} parameter(0)
+param.2 = bf16[8,128,1024]{2,1,0} parameter(1)
+reduce-scatter.1 = bf16[8,64,1024]{2,1,0} reduce-scatter(param.1), channel_id=8, replica_groups={{0,1},{2,3},{4,5},{6,7}}, use_global_device_ids=true, dimensions={1}, to_apply=add
+all-gather.1 = bf16[8,128,1024]{2,1,0} all-gather(reduce-scatter.1), channel_id=5, replica_groups={{0,1},{2,3},{4,5},{6,7}}, dimensions={1}, use_global_device_ids=true
+reduce-scatter.2 = bf16[8,64,1024]{2,1,0} reduce-scatter(param.2), channel_id=9, replica_groups={{0,1},{2,3},{4,5},{6,7}}, use_global_device_ids=true, dimensions={1}, to_apply=add
+all-gather.2 = bf16[8,128,1024]{2,1,0} all-gather(reduce-scatter.2), channel_id=5, replica_groups={{0,1},{2,3},{4,5},{6,7}}, dimensions={1}, use_global_device_ids=true
+add.1 = bf16[8,128,1024]{2,1,0} add(all-gather.1, all-gather.2)
+}
+)";
+
+  TF_ASSERT_OK_AND_ASSIGN(auto module, RunPass(hlo_string,
+                                               /*num_replicas=*/8,
+                                               /*num_partitions=*/1,
+                                               /*expect_change=*/true));
+  EXPECT_EQ(AllGatherCount(module), 0);
+  EXPECT_EQ(ReduceScatterCount(module), 0);
+  EXPECT_EQ(AllReduceCount(module), 2);
+}
+
+TEST_F(ReduceScatterAllGatherCombinerTest, ReduceScatterWithMoreThanOneUser) {
+  /* In the following pattern only the left side should be reduced as the right
+  reduce scatter has two children
+
+  Param              Param            Param
+    |                  |                |
+  ReduceScatter   ReduceScatter   ReduceScatter
+  |               |          \     /        |
+  AllGather       AllGather    Add      AllGather
```

Comment:
May be the test can be more targeted with just the pattern on the right (RS used 2 times) and expect no change?

Reply:
done.